### PR TITLE
Checkpoint Soup: average weights from multiple training stages

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

**Model Soups** (Wortsman et al., ICML 2022) demonstrated that averaging model weights from different training stages or configurations produces models that are more robust and accurate than any individual checkpoint. The key insight: different stages of training explore different basins in weight space, and their average lies in a flatter, more generalizable region.

**Our setup is ideal for this:** We already train with EMA (decay=0.999) and save the best EMA checkpoint. But we only use ONE checkpoint at inference. By saving 3 EMA checkpoints at different training stages (e.g., best_val_loss, epoch 120, epoch 140) and averaging their weights at inference, we can create a "soup" model that combines diverse representations.

**Why this should improve OOD:** Each checkpoint has different strengths — early checkpoints are more conservative (better OOD), late checkpoints are more specialized (better in-dist). Weight averaging produces a model that balances both. Wortsman et al. showed 1-3% improvement on OOD benchmarks.

**Zero training cost.** This only changes the inference/evaluation procedure — the training is identical to baseline. We just save extra checkpoints and average weights after training.

### Step 1: Save multiple checkpoints during training

Modify the training loop to save EMA checkpoints at 3 points:
```python
parser.add_argument('--checkpoint_soup', action='store_true',
                    help='Save multiple checkpoints for model soup averaging')
parser.add_argument('--soup_epochs', type=str, default='100,125',
                    help='Comma-separated epochs at which to save extra soup checkpoints')
```

During training, after updating the EMA:
```python
if args.checkpoint_soup:
    soup_epochs = [int(e) for e in args.soup_epochs.split(',')]
    if epoch in soup_epochs:
        torch.save(ema_model.state_dict(), f'soup_checkpoint_epoch{epoch}.pt')
```

The final EMA model is also saved as usual.

### Step 2: Average weights at evaluation

After training completes, before validation:
```python
if args.checkpoint_soup:
    soup_epochs = [int(e) for e in args.soup_epochs.split(',')]
    state_dicts = []
    for ep in soup_epochs:
        ckpt_path = f'soup_checkpoint_epoch{ep}.pt'
        if os.path.exists(ckpt_path):
            state_dicts.append(torch.load(ckpt_path, map_location='cuda'))
    # Add the final EMA model
    state_dicts.append(ema_model.state_dict())
    
    # Average all saved state dicts
    avg_state = {}
    for key in state_dicts[0]:
        avg_state[key] = sum(sd[key].float() for sd in state_dicts) / len(state_dicts)
    
    # Load averaged weights into model for evaluation
    ema_model.load_state_dict(avg_state)
    print(f"Checkpoint soup: averaged {len(state_dicts)} checkpoints")
```

### Step 3: Run 2 seeds with soup at epochs 100, 125

```bash
cd cfd_tandemfoil && python train.py \
  --agent frieren --wandb_name "frieren/soup-100-125-s42" --seed 42 \
  --wandb_group checkpoint-soup \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling \
  --checkpoint_soup --soup_epochs 100,125

# Repeat for seed 73: --seed 73 --wandb_name "frieren/soup-100-125-s73"
```

**Important:** The training itself is IDENTICAL to baseline — only the final evaluation uses the averaged weights. The W&B metrics logged during training (per-epoch val) will show baseline-like performance. The soup metrics only appear in the FINAL evaluation after weight averaging.

**Log both:** Log the pre-soup best_val metrics AND the post-soup metrics so we can compare the effect of averaging.

## Baseline

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in | 11.742 | < 11.742 |
| p_oodc | 7.643 | < 7.643 |
| p_tan | 27.874 | < 27.874 |
| p_re | 6.419 | < 6.419 |

- **Baseline W&B runs:** k5qwvce4 (seed 42), 7oa5xfhi (seed 73)
- **val/loss baseline:** ~0.37

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Pre-soup (standard EMA best checkpoint)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| p_in | 11.4 | 11.9 | 11.65 | 11.742 | **-0.8%** ✅ |
| p_tan | 28.4 | 29.4 | 28.90 | 27.874 | +3.7% ❌ |
| p_oodc | 7.8 | 7.6 | 7.70 | 7.643 | +0.7% |
| p_re | 6.4 | 6.4 | 6.40 | 6.419 | **-0.3%** ✅ |
| val/loss | 0.3796 | 0.3769 | 0.378 | ~0.37 | +2.2% |

### Post-soup (3-checkpoint average: epoch 100 + epoch 125 + best EMA)

| Metric | Seed 42 | Seed 73 | 2-seed avg | vs Pre-soup |
|--------|---------|---------|------------|-------------|
| p_in | 12.1 | 12.5 | 12.30 | **+5.6% ❌** |
| p_tan | 28.2 | 29.3 | 28.75 | -0.5% |
| p_oodc | 8.1 | 7.9 | 8.00 | **+3.9% ❌** |
| p_re | 6.7 | 6.8 | 6.75 | **+5.5% ❌** |

### Analysis

**The checkpoint soup hurt performance significantly.** Weight averaging across epochs 100, 125, and best (~140) degraded surface pressure prediction on 3 of 4 splits:

- **p_in worsened by +5.6%** — the most dramatic regression
- **p_oodc worsened by +3.9%**
- **p_re worsened by +5.5%**
- **p_tan improved marginally by -0.5%** — likely noise

**Why it didn't work:** The Model Soups paper showed improvements when averaging models trained with *different hyperparameters* (different learning rates, augmentation). Here, we averaged checkpoints from *the same training run* at different stages. The EMA at epoch 100 is substantially undertrained compared to the final EMA — it hasn't had ~40 more epochs of refinement. Weight averaging with an undertrained snapshot drags the final model toward worse representations, especially for the fine-grained surface pressure predictions this model specializes in.

The ema_start_epoch was adjusted from 140 → 94 to accommodate the experiment (giving 6 warmup epochs before the first soup checkpoint at 100). This means the EMA was collecting updates from much earlier, which itself may slightly differ from baseline behavior — though the pre-soup metrics are within normal seed variation of baseline.

**Note:** The earlier EMA start (94 vs 140) for soup compatibility caused the pre-soup p_tan to regress (+3.7% vs baseline). This is likely because the EMA was accumulating from a less-trained model, diluting the final EMA quality compared to the default ema_start_epoch=140.

### Training details
- **Command:** `python train.py --agent frieren --wandb_name "frieren/soup-100-125-s{42,73}" --seed {42,73} --wandb_group checkpoint-soup --checkpoint_soup --soup_epochs 100,125` (plus all baseline flags)
- **W&B runs:** v5af1m2c (seed 42), pooovezz (seed 73)
- **Peak memory:** 46.1–46.4 GB
- **Total epochs:** 140–141 (180 min timeout)
- **Soup checkpoints saved:** epoch 100, epoch 125, plus best model (~epoch 140)

### Suggested follow-ups

1. **Greedy soup** — instead of uniform averaging, only include checkpoints that improve val_loss when added to the average (Wortsman et al.'s greedy selection method).
2. **Late-stage soup** — average only epochs near the end (e.g. 130, 135, best) where the model is fully converged, avoiding the undertrained epoch-100 checkpoint.
3. **Multi-run soup** — train 3+ independent runs with different seeds and average their final checkpoints (the original Model Soups setup). This is the setting where soup was shown to help most.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2334 because Checkpoint soup (averaging epochs 100+125+best) worsened 3/4 metrics: p_in +5.6%, p_oodc +3.9%, p_re +5.5% vs pre-soup checkpoint. Averaging undertrained checkpoints (epoch 100) with the final model drags performance toward worse representations. Original Model Soups works with different hyperparameters, not different epochs of the same run. Weight averaging from same-run checkpoints is exhausted. W&B: v5af1m2c (s42), pooovezz (s73)..
